### PR TITLE
fix(engine): release trigger provider guards before await

### PR DESCRIPTION
**Pull request by guibeira**
## Summary

Trigger registry fix related to #2178, kept in draft for review and CI validation.

- Uses snapshots with Arc-shared registrators to release DashMap guards before asynchronous registration, replay, and removal calls.
- Separates pending-binding collection from provider resolution, avoiding cross-map lookups during iteration.
- Detects provider-generation replacement during pending replay and forwards activation to the current provider before publishing it.
- Adds regression tests for guards held across await, generation replacement, and concurrent reconnection.

## Initial local validation

- The `replay_releases_provider_shard_before_await` regression failed on the original code and passed after the fix.
- `cargo test -p iii --lib trigger::tests --locked`: 79 tests passed for the initial fix. Subsequent validation results are documented in the review discussions.
- The concurrent scenario runs 20 rounds with 8 providers, 48 types, and 576 bindings, checking pending-binding recovery and delivery counts.
- `cargo fmt --all` and `git diff --check` completed successfully.

## Limitations

Validation covers Rust components. No end-to-end validation of the patched binary with WebSockets, RPC, and shutdown was performed; the tests should not be interpreted as that evidence. No active project was restarted. Remote CI results are tracked separately in the PR checks.

Refs #2178.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved trigger provider handling during asynchronous operations.
  - Improved reliability when providers reconnect or are replaced while pending triggers are replayed.
  - Prevented stale provider registrations from being activated after newer providers become current.
  - Improved cleanup and consistency of in-flight trigger deliveries during worker or provider changes.
  - Added operation deadlines to prevent trigger delivery and acknowledgement waits from hanging indefinitely.
  - Ensured shutdown signal handlers are ready before shutdown readiness is reported.
- **Tests**
  - Expanded coverage for trigger lifecycle, delivery timeouts, provider replacement, and shutdown signal ordering.

### CODE REVIEW

**Comment by guibeira** [comment]
@coderabbitai review

File: engine/src/trigger.rs [hunk 1]
```diff
@@ -781,7 +824,7 @@ impl TriggerRegistry {
 
         let Some(trigger_type) = resolved
             .as_ref()
-            .and_then(|key| self.trigger_types.get(key))
+            .and_then(|key| self.provider_snapshot(key))
```
[the remark this answers is not in the document]
Comment (guibeira) [reply engine/src/trigger.rs]: Addressed in e43e15d14. Binding-scoped async lifecycle gates now serialize actual provider delivery/replay/detach, while a short synchronous publication fence coordinates provider replacement with final binding publication and gate release. No DashMap guard crosses an await. Replacement skips busy binding gates; their holders reconcile against the current provider, rather than blocking the provider acknowledgement loop. Replay batches re-read bindings after acquiring their gate, so stale candidates cannot resurrect a removed binding.

Accepted deliveries are tracked by provider generation for teardown. The same settlement path covers direct registration, failover, pending recovery and re-homing. Rejected re-registration preserves real prior deliveries (including rejected route/type changes); disconnect cleanup selects relevant bindings instead of waiting on unrelated slow registrations; empty lifecycle gates are reclaimed without splitting queued callers. Cancelled detach retains already-recorded deliveries.

Validation: `cargo fmt --all && timeout 180s cargo test -p iii --lib trigger::tests --locked && git diff --check` passed, with 88 selected tests. New deterministic tests check actual provider-side delivery/detach during registration, failover and unregister replacement races, pending/re-home recovery with owner disconnect, rejection races, unrelated-cleanup liveness, gate churn and cancelled detach. Existing shard-release and concurrent-reconnect regressions remain passing.

Remaining scope limits: this is component validation, not WebSocket/RPC/shutdown E2E. Provider detach failures remain best-effort during failover; failed deliveries are retained, but there is no dedicated owner-disconnect retry queue. Registration/replay cancellation after an external side effect but before acknowledgement is not transactional. The PR remains draft pending fresh CI and your follow-up review; this thread is not marked resolved yet.

**Comment by guibeira** [comment]
@coderabbitai full review

CI is now green for e43e15d14cc2f3ebe66c07f89cdfc6db9ad88403. Please review the complete updated diff, including the binding lifecycle synchronization and new regression tests, and refresh the pre-merge checks/docstring coverage. The original race finding was confirmed resolved in the review thread; this requests a full follow-up review of the current draft, not only that thread.

File: engine/src/trigger.rs [hunk 2]
```diff
@@ -398,346 +620,192 @@ impl TriggerRegistry {
...
-                        trigger.id.purple(),
-                        trigger.trigger_type.purple(),
-                        trigger.function_id.purple(),
-                    ),
                 }
-                self.pending_triggers.insert(trigger.id.clone(), trigger);
+                self.settle_binding(state, trigger, None).await;
             }
         }
     }
```
[the remark this answers is not in the document]
Comment (guibeira) [reply engine/src/trigger.rs:832]: Addressed in 6911ac845. Every register/replay/unregister future awaited at the binding lifecycle boundary now shares one absolute deadline per acquired gate, using the existing ten-second registration policy (a shared compile-time constant, not a new runtime configuration setting). Reconciliation and multiple detaches do not reset that budget. WorkerConnection send/ack use the same deadline, preserving fail-open behavior after successful delivery without an ack. A cancellation guard cleans closed ack waiters without deleting a replacement's open waiter.

Accepted and uncertain deliveries remain tracked until successful detach. Explicit unregister can retry retained deliveries even after owner cleanup removed both live/pending entries. Documentation was added to deadline helpers and touched test helpers; please refresh docstring coverage on follow-up review.

Validation: 92 selected trigger tests and 16 worker-connection tests passed, plus cargo fmt --all and git diff --check. New paused-time tests use actual WorkerConnection channels held full to verify bounded registration/replay/owner detach, gate release, ack cleanup, and subsequent explicit teardown. Tests also cover cancellation and preserved fail-open ack semantics.

Limits: deadlines are cooperative and begin after gate acquisition; this is a per-binding budget, not a whole-worker cleanup deadline. Permanent provider failures require explicit retry; no background retry or transactional external rollback is claimed. The PR remains draft pending fresh CI and follow-up review.

**Comment by guibeira** [comment]
CI follow-up: Engine Coverage on 6911ac845 failed in `test_ingest_metrics_multiple_resources` (global metric count 3 vs expected 2; 1787 passed, 1 failed, 1 ignored). Fixed in 46ea9173a by giving five OTLP ingestion edge-case tests owned storage through an extracted ingestion helper, retaining their exact-count and conversion assertions.

Investigation found a concrete non-serialized test that clears/writes the same process-global metrics store; serial annotations on the ingestion tests do not exclude that writer. The failed log does not identify the third metric, so this is evidence of a real interference path, not proof of which writer ran in that CI schedule. A deterministic regression reproduces the extra-entry interleaving and verifies isolated storage excludes it; another preserves malformed-batch/no-partial-write behavior.

Validation: 32 ingestion-selected tests passed; the full local Engine library suite passed on its first run (1790 passed, 0 failed, 1 ignored). Formatting and diff checks passed. Coverage-instrumented CI is now rerunning; the PR remains draft.

**Comment by guibeira** [comment]
CI follow-up: Worker Tests (VM) - macos failed on 46ea9173a in `sandbox_signal_exit_writes_no_engine_gone_breadcrumb` (SIGTERM terminated the child instead of graceful exit; 1383 tests passed, 1 failed). Fixed in ca5924d66.

Verified ordering bug: ExitWatch::wait installed signal handlers lazily inside select branches. Randomized polling could log `parent exit-watch armed` before installing SIGTERM handling; the test immediately signals after observing that log. Unix handlers are now installed synchronously before polling any watch. The strengthened regression verifies signal-readiness precedes parent-watch readiness, then sends SIGTERM immediately—no added sleep or weakened assertion.

Local Linux validation passed: all 6 sandbox daemon integration tests, 5 daemon_exit unit tests, and the 6 integration tests with the CI-matching `integration-vm` feature. Formatting/diff checks passed. macOS behavior still requires CI confirmation. Signal delivery earlier than ExitWatch startup is outside this readiness contract. PR remains draft pending CI and full follow-up review.

**Comment by guibeira** [comment]
@coderabbitai full review

All CI checks are now green for ca5924d6698808bd9b72aef23b83c4aa8d813a62, including Engine Coverage and Worker Tests (VM) - macos. Please review the complete current diff and refresh pre-merge/docstring coverage checks. Both earlier code findings were confirmed resolved in their threads. Since the last full review, changes include bounded trigger lifecycle operations, OTLP ingestion test storage isolation, and shutdown signal installation before exit-watch readiness. The PR remains draft pending this follow-up review.

**Review by sergiofilhowz — COMMENTED** [review]
## Adversarial review: 10 failing tests, 4 of them regressions vs `main`

The core fix looks right: no DashMap guard is held across `.await` anymore, the 92 existing `trigger::tests` pass, and CI is green. The new lifecycle layer (gates / `deliveries` / `uncertain` / `reconcile_existing`) introduces regressions, though, mostly around providers whose connection is already dead.

I wrote 10 targeted tests. **All 10 fail on this branch (`9ed09e00d`).** The 5 that also compile on `main` (`f2bc2b2cd`) were run there as a baseline, and 4 of them pass on `main`.

| Test | PR | main | Verdict |
|---|---|---|---|
| R1 provider restart → consumer unregisters | ❌ | ✅ | **regression (severe)** |
| R2 unregister a parked intent while the provider is down | ❌ | ✅ | **regression (severe)** |
| R3 home provider leaves → fallback → unregister | ❌ | ✅ | **regression (severe)** |
| R4 same connection re-announces its trigger type | ❌ | ✅ | regression |
| R5 route change leaves the old provider's job running | ❌ | ❌ | pre-existing |
| R6 / R7 gate + dead `WorkerConnection` leak | ❌ | n/a | PR-only |
| R8 / R9 / R10 cleanup time / blocking | ❌ | n/a | PR-only |

The tests use a real `WorkerConnection` and drop its receiver to model the actual teardown order in `handle_worker` (`engine/mod.rs:2629`): `writer.abort()` runs **before** `cleanup_worker`, so every later send to a departed provider fails with `channel closed`.

---

### 🔴 1. Bindings become impossible to unregister after a provider restart or departure (R1, R2, R3)

- **Where the dead delivery is kept:** `detach_deliveries` (`trigger.rs:468`) keeps a delivery whose detach failed. For a dead connection that detach can never succeed, so the delivery stays forever.
- **Where the binding comes back:** when the consumer unregisters, `unregister_trigger` (`trigger.rs:996`) calls `reconcile_existing` → `settle_binding`. That path **replays the binding to the current provider** it just detached from, then returns `Err`.

Observed results:
- **R1** (the #2178 reconnect flow): the provider receives `["-t", "+t"]`, the call returns `Err(failed to send unregister trigger message through worker channel: channel closed)`, and the binding stays live.
- **R2:** the parked intent is re-parked, then silently re-activated when the provider reconnects (`replayed_on_return=["+t"]`).
- **R3:** happens with no restart at all. Fallback provider events: `["+t", "-t", "+t"]`.

The router drops the result (`let _ = … unregister_trigger(…)`, `engine/mod.rs:1602`). So the worker believes the trigger is gone, and it keeps firing. Every retry repeats the detach + replay churn.

### 🟠 2. Unbounded leak per binding per provider restart (R6, R7)

- `LifecycleLease::drop` only reclaims a gate when `deliveries` is empty, so failed dead-generation deliveries pin the gate forever. Each one holds a clone of the dead `WorkerConnection`.
- **R6:** after the consumer disconnects, both maps are empty but the gate is retained (`Some(1)` delivery).
- **R7:** after 5 provider restarts a single binding retains **6** provider generations. This grows linearly with restarts × bindings.

### 🟠 3. The same connection re-announcing its type counts as a new generation (R4)

Generation identity is `Arc::ptr_eq`, but the router builds a fresh `Box::new(worker.clone())` for every `RegisterTriggerType`. So the same socket receives `["-t", "+t"]`, leaving a window with no job (for example an HTTP route gap or a cron reset). `main` sends only `+t`.

### 🟡 4. The 10 s budget applies per binding, not per operation (R8)

A consumer with 30 bindings on a stalled provider blocks `unregister_worker` (and therefore `cleanup_worker`) for **300 s** of virtual time. That's much better than `main`, which blocks forever, but shutdown/cleanup is still delayed linearly in the number of bindings.

### 🟡 5. `disconnect_does_not_wait_for_unrelated_busy_binding` passes only because it uses explicit namespaces (R9)

With **resolved** bindings (what workers send by default), `affected()` (`trigger.rs:751`) matches every binding of that type whenever the departing provider lives in `default`. An in-flight snapshot even carries `provider_namespace = "default"`. As a result, disconnecting the DEFAULT provider waits on the gate of a binding that is being delivered to an unrelated `shop` provider.

### 🟡 6. Head-of-line blocking on a provider's read loop (R10)

1. An ownerless registration (`engine::register_trigger`) holds the gate while it waits for provider P's ack.
2. P's read loop then handles an `UnregisterTrigger` for the same id and blocks on that gate.
3. P's read loop is the only thing that could read the ack that would release the gate.

It resolves only at the 10 s fail-open deadline, and every message on P's connection is stalled meanwhile. This isn't a correctness regression (on `main` the unregister is simply lost and the binding orphaned), but it's a new stall.

### ⚪ 7. Other notes

- **R5 (pre-existing):** re-registering an id onto another provider never detaches the previous provider, so both fire. Now that `deliveries` are tracked, this is easy to fix on successful publication.
- **Operator logs removed:** `[REGISTERED] Trigger Type …`, `[RE-HOMED]`, `[DISABLED] … will be re-registered automatically`, `recovered from pending`, `Unregistering trigger: …` and `Trigger provider unreachable; bind parked as pending` are all gone.
- **Inconsistent error classification:** with a closed channel, `WorkerConnection::replay_trigger` returns a plain `anyhow` error, while `register_trigger` returns `RegistratorUnavailable`. The two feed the `uncertain` bookkeeping differently.
- **Unrelated changes:** the `daemon_exit` and metrics-test changes aren't related to #2178, and splitting them out would ease review and bisecting. `test_ingest_metrics_owned_storage_excludes_interleaved_writer` spawns and immediately joins a thread, so nothing actually interleaves.
- **Public API change:** the `TriggerType.registrator` field changed from `Box` to `Arc`.

### Suggested fixes

1. Treat a closed provider channel as a definitive detach, since the jobs die with the connection. This could be a typed `RegistratorGone` error or a `TriggerRegistrator::is_closed()`. `unregister_worker` should also drop the deliveries owned by the departing worker.
2. In `unregister_trigger`, never replay once the current generation has been detached. Stale-generation failures must not block removal; keep them, if at all, in a retry-only list that doesn't pin the gate.
3. Identify a generation by its connection (`worker_id`), not by the `Arc` allocation.
4. On successful publication, detach deliveries left on other providers (fixes R5).
5. Use one deadline per operation, or detach concurrently. In `unregister_worker`, skip busy gates of bindings the departing worker doesn't own, the same way `replay_candidate` does (the holder already re-resolves under the fence).
6. Restore the removed lifecycle logs.

### Reproduce

Append the two modules below to the end of `engine/src/trigger.rs`, then run:

```
SKIP_UI_BUILD=1 cargo test -p iii --lib review_2198
```

(`review_2198_portable` also compiles on `main`; `review_2198_pr_only` relies on this PR's internals.)

<details>
<summary>Test code (R1–R10)</summary>

```rust
// Real teardown order (engine/src/engine/mod.rs, handle_worker):
//     writer.abort();                 // drops the channel Receiver
//     self.cleanup_worker(&worker)    // -> trigger_registry.unregister_worker
// so every send to a departed provider fails with "channel closed".
#[cfg(test)]
mod review_2198_portable {
    use super::*;
    use crate::engine::Outbound;
    use crate::protocol::{DEFAULT_NAMESPACE, Message};
    use crate::worker_connections::WorkerConnection;
    use std::collections::HashSet;
    use std::sync::Arc;

    #[derive(Default)]
    pub(super) struct Probe {
        pub live: std::sync::Mutex<HashSet<String>>,
        pub events: std::sync::Mutex<Vec<String>>,
    }

    impl TriggerRegistrator for Arc<Probe> {
        fn register_trigger(
            &self,
            trigger: Trigger,
        ) -> Pin<Box<dyn Future<Output = Result<(), anyhow::Error>> + Send + '_>> {
            Box::pin(async move {
                self.live.lock().unwrap().insert(trigger.id.clone());
                self.events.lock().unwrap().push(format!("+{}", trigger.id));
                Ok(())
            })
        }

        fn unregister_trigger(
            &self,
            trigger: Trigger,
        ) -> Pin<Box<dyn Future<Output = Result<(), anyhow::Error>> + Send + '_>> {
            Box::pin(async move {
                self.live.lock().unwrap().remove(&trigger.id);
                self.events.lock().unwrap().push(format!("-{}", trigger.id));
                Ok(())
            })
        }
    }

    pub(super) fn base_trigger(id: &str) -> Trigger {
        Trigger {
            id: id.to_string(),
            trigger_type: "evt".to_string(),
            function_id: format!("fn_{id}"),
            config: serde_json::json!({}),
            worker_id: None,
            metadata: None,
            namespace: DEFAULT_NAMESPACE.to_string(),
            trigger_namespace: None,
            home_namespace: DEFAULT_NAMESPACE.to_string(),
            provider_namespace: DEFAULT_NAMESPACE.to_string(),
        }
    }

    /// A binding sent by a consumer worker (connection-owned, fire-and-forget).
    pub(super) fn consumer_trigger(id: &str, owner: Uuid) -> Trigger {
        Trigger {
            worker_id: Some(owner),
            ..base_trigger(id)
        }
    }

    pub(super) fn worker_provider(
        capacity: usize,
    ) -> (WorkerConnection, tokio::sync::mpsc::Receiver<Outbound>) {
        let (tx, rx) = tokio::sync::mpsc::channel(capacity);
        (WorkerConnection::new(tx), rx)
    }

    /// Mirrors router_msg's RegisterTriggerType: a fresh Box per message.
    pub(super) async fn install_worker_provider(
        registry: &TriggerRegistry,
        namespace: &str,
        connection: &WorkerConnection,
    ) {
        registry
            .register_trigger_type(TriggerType::new_ns(
                namespace,
                "evt",
                "worker provider",
                Box::new(connection.clone()),
                Some(connection.id),
            ))
            .await
            .unwrap();
    }

    pub(super) fn drain(rx: &mut tokio::sync::mpsc::Receiver<Outbound>) -> Vec<String> {
        let mut out = Vec::new();
        while let Ok(msg) = rx.try_recv() {
            out.push(match msg {
                Outbound::Protocol(Message::RegisterTrigger { id, .. }) => format!("+{id}"),
                Outbound::Protocol(Message::UnregisterTrigger { id, .. }) => format!("-{id}"),
                _ => "?".to_string(),
            });
        }
        out
    }

    /// Provider P1 dies (writer aborted, then cleanup) and reconnects as P2.
    /// Returns P2 and its receiver with the replay already drained.
    pub(super) async fn provider_restart_with_binding(
        registry: &TriggerRegistry,
        consumer: Uuid,
    ) -> (WorkerConnection, tokio::sync::mpsc::Receiver<Outbound>) {
        let (p1, rx1) = worker_provider(8);
        install_worker_provider(registry, DEFAULT_NAMESPACE, &p1).await;
        registry
            .register_trigger(consumer_trigger("t", consumer))
            .await
            .unwrap();
        drop(rx1);
        registry.unregister_worker(&p1.id).await;
        assert!(registry.pending_triggers.contains_key("t"));

        let (p2, mut rx2) = worker_provider(8);
        install_worker_provider(registry, DEFAULT_NAMESPACE, &p2).await;
        assert!(registry.triggers.contains_key("t"));
        assert_eq!(drain(&mut rx2), vec!["+t"]);
        (p2, rx2)
    }

    /// R1 — provider restart (the exact #2178 flow), then the consumer
    /// unregisters. Expected: Ok(true), binding gone, P2 told "-t" only.
    #[tokio::test]
    async fn r1_unregister_after_provider_restart_really_removes_binding() {
        let registry = TriggerRegistry::new();
        let (_p2, mut rx2) = provider_restart_with_binding(&registry, Uuid::new_v4()).await;

        let result = registry.unregister_trigger("t".into(), None).await;
        let sent = drain(&mut rx2);
        assert!(
            result.is_ok() && !registry.triggers.contains_key("t") && sent == vec!["-t"],
            "unregister={result:?} still_live={} provider_saw={sent:?}",
            registry.triggers.contains_key("t")
        );
    }

    /// R2 — the provider is down, the binding is parked; the consumer
    /// unregisters it. Expected: Ok(true) and NOTHING replayed when the
    /// provider comes back.
    #[tokio::test]
    async fn r2_unregister_parked_intent_while_provider_is_down() {
        let registry = TriggerRegistry::new();
        let (p1, rx1) = worker_provider(8);
        install_worker_provider(&registry, DEFAULT_NAMESPACE, &p1).await;
        registry
            .register_trigger(consumer_trigger("t", Uuid::new_v4()))
            .await
            .unwrap();
        drop(rx1);
        registry.unregister_worker(&p1.id).await;
        assert!(registry.pending_triggers.contains_key("t"));

        let result = registry.unregister_trigger("t".into(), None).await;
        let still_parked = registry.pending_triggers.contains_key("t");

        let (p2, mut rx2) = worker_provider(8);
        install_worker_provider(&registry, DEFAULT_NAMESPACE, &p2).await;
        let replayed = drain(&mut rx2);
        assert!(
            result.is_ok()
                && !still_parked
                && replayed.is_empty()
                && !registry.triggers.contains_key("t"),
            "unregister={result:?} still_parked={still_parked} replayed_on_return={replayed:?} live_again={}",
            registry.triggers.contains_key("t")
        );
    }

    /// R3 — no restart at all: a home provider leaves, the binding falls back
    /// to the DEFAULT provider, then the consumer unregisters it.
    #[tokio::test]
    async fn r3_unregister_after_home_provider_left_for_fallback() {
        let registry = TriggerRegistry::new();
        let fallback = Arc::new(Probe::default());
        registry
            .register_trigger_type(TriggerType::new_ns(
                DEFAULT_NAMESPACE,
                "evt",
                "fallback",
                Box::new(Arc::clone(&fallback)),
                None,
            ))
            .await
            .unwrap();
        let (home, rx_home) = worker_provider(8);
        install_worker_provider(&registry, "shop", &home).await;
        let binding = Trigger {
            home_namespace: "shop".into(),
            ..consumer_trigger("t", Uuid::new_v4())
        };
        registry.register_trigger(binding).await.unwrap();
        assert_eq!(registry.triggers.get("t").unwrap().provider_namespace, "shop");

        drop(rx_home);
        registry.unregister_worker(&home.id).await;
        assert!(fallback.live.lock().unwrap().contains("t"));

        let result = registry.unregister_trigger("t".into(), None).await;
        let events = fallback.events.lock().unwrap().clone();
        assert!(
            result.is_ok() && fallback.live.lock().unwrap().is_empty(),
            "unregister={result:?} fallback_still_runs={} fallback_events={events:?}",
            fallback.live.lock().unwrap().contains("t")
        );
    }

    /// R4 — the SAME connection re-announcing its trigger type (router builds
    /// a fresh Box each time) must not tear bindings down. main replays "+t";
    /// a detach first opens a window where the provider has no job.
    #[tokio::test]
    async fn r4_same_connection_reregistration_does_not_detach() {
        let registry = TriggerRegistry::new();
        let (p, mut rx) = worker_provider(8);
        install_worker_provider(&registry, DEFAULT_NAMESPACE, &p).await;
        registry
            .register_trigger(consumer_trigger("t", Uuid::new_v4()))
            .await
            .unwrap();
        drain(&mut rx);
        install_worker_provider(&registry, DEFAULT_NAMESPACE, &p).await;
        let sent = drain(&mut rx);
        assert!(!sent.contains(&"-t".to_string()), "same connection saw {sent:?}");
    }

    /// R5 — re-registering an id onto another provider (route change) must
    /// retire the job on the previous provider; otherwise both fire.
    /// (Pre-existing on main, where the old job is even untracked.)
    #[tokio::test]
    async fn r5_route_change_detaches_previous_provider() {
        let registry = TriggerRegistry::new();
        let old = Arc::new(Probe::default());
        let other = Arc::new(Probe::default());
        for (ns, probe) in [(DEFAULT_NAMESPACE, &old), ("other", &other)] {
            registry
                .register_trigger_type(TriggerType::new_ns(
                    ns,
                    "evt",
                    "probe",
                    Box::new(Arc::clone(probe)),
                    None,
                ))
                .await
                .unwrap();
        }
        registry.register_trigger(base_trigger("t")).await.unwrap();
        let moved = Trigger {
            trigger_namespace: Some("other".into()),
            ..base_trigger("t")
        };
        registry.register_trigger(moved).await.unwrap();
        assert!(other.live.lock().unwrap().contains("t"));
        assert!(
            old.live.lock().unwrap().is_empty(),
            "old provider still runs superseded binding: {:?}",
            old.events.lock().unwrap()
        );
    }
}

// Needs PR-only internals (lifecycles, TRIGGER_OPERATION_TIMEOUT).
#[cfg(test)]
mod review_2198_pr_only {
    use super::review_2198_portable::*;
    use super::*;

    fn retained_deliveries(registry: &TriggerRegistry, id: &str) -> Option<usize> {
        registry.lifecycles.get(id).map(|gate| {
            gate.try_lock()
                .map(|s| s.deliveries.len() + s.uncertain.len())
                .unwrap_or(usize::MAX)
        })
    }

    /// R6 — the consumer disconnects after a provider restart. Both maps are
    /// cleared, but the gate (holding the dead WorkerConnection) must go too.
    #[tokio::test]
    async fn r6_owner_disconnect_after_provider_restart_reclaims_gate() {
        let registry = TriggerRegistry::new();
        let consumer = Uuid::new_v4();
        let (_p2, _rx2) = provider_restart_with_binding(&registry, consumer).await;
        registry.unregister_worker(&consumer).await;
        assert!(registry.triggers.is_empty());
        assert!(registry.pending_triggers.is_empty());
        assert!(
            registry.lifecycles.is_empty(),
            "gate leaked with {:?} retained deliveries",
            retained_deliveries(&registry, "t")
        );
    }

    /// R7 — every provider restart appends one more dead generation to the
    /// binding's delivery list (each one retains a WorkerConnection clone).
    #[tokio::test]
    async fn r7_retained_generations_do_not_grow_with_restarts() {
        let registry = TriggerRegistry::new();
        let (mut provider, mut rx) = worker_provider(8);
        install_worker_provider(&registry, crate::protocol::DEFAULT_NAMESPACE, &provider).await;
        registry
            .register_trigger(consumer_trigger("t", Uuid::new_v4()))
            .await
            .unwrap();
        for _ in 0..5 {
            drop(rx);
            registry.unregister_worker(&provider.id).await;
            let (next, next_rx) = worker_provider(8);
            install_worker_provider(&registry, crate::protocol::DEFAULT_NAMESPACE, &next).await;
            provider = next;
            rx = next_rx;
        }
        let retained = retained_deliveries(&registry, "t");
        assert!(
            retained <= Some(1),
            "after 5 restarts the binding retains {retained:?} provider generations"
        );
    }

    /// Registration succeeds; detach never completes (e.g. provider stopped
    /// draining its socket).
    struct StuckDetach;

    impl TriggerRegistrator for StuckDetach {
        fn register_trigger(
            &self,
            _trigger: Trigger,
        ) -> Pin<Box<dyn Future<Output = Result<(), anyhow::Error>> + Send + '_>> {
            Box::pin(async { Ok(()) })
        }

        fn unregister_trigger(
            &self,
            _trigger: Trigger,
        ) -> Pin<Box<dyn Future<Output = Result<(), anyhow::Error>> + Send + '_>> {
            Box::pin(std::future::pending())
        }
    }

    /// R8 — the 10 s budget is per binding, not per operation: a consumer with
    /// N bindings on a stalled provider blocks cleanup_worker for N x 10 s.
    #[tokio::test(start_paused = true)]
    async fn r8_owner_cleanup_is_not_linear_in_bindings() {
        const N: u32 = 30;
        let registry = TriggerRegistry::new();
        registry
            .register_trigger_type(TriggerType::new(
                "evt",
                "stuck",
                Box::new(StuckDetach),
                Some(Uuid::new_v4()),
            ))
            .await
            .unwrap();
        let owner = Uuid::new_v4();
        for i in 0..N {
            registry
                .register_trigger(consumer_trigger(&format!("t{i}"), owner))
                .await
                .unwrap();
        }
        let start = tokio::time::Instant::now();
        registry.unregister_worker(&owner).await;
        let elapsed = tokio::time::Instant::now() - start;
        assert!(
            elapsed <= TRIGGER_OPERATION_TIMEOUT,
            "owner cleanup took {elapsed:?} (virtual) for {N} bindings"
        );
    }

    #[derive(Default)]
    struct PausingProbe {
        pause: std::sync::atomic::AtomicBool,
        resume: tokio::sync::Notify,
    }

    impl TriggerRegistrator for std::sync::Arc<PausingProbe> {
        fn register_trigger(
            &self,
            _trigger: Trigger,
        ) -> Pin<Box<dyn Future<Output = Result<(), anyhow::Error>> + Send + '_>> {
            Box::pin(async move {
                if self.pause.swap(false, std::sync::atomic::Ordering::SeqCst) {
                    self.resume.notified().await;
                }
                Ok(())
            })
        }

        fn unregister_trigger(
            &self,
            _trigger: Trigger,
        ) -> Pin<Box<dyn Future<Output = Result<(), anyhow::Error>> + Send + '_>> {
            Box::pin(async { Ok(()) })
        }
    }

    /// R9 — variant of `disconnect_does_not_wait_for_unrelated_busy_binding`
    /// using RESOLVED bindings (what workers send by default) instead of
    /// explicit namespaces.
    #[tokio::test]
    async fn r9_default_provider_disconnect_waits_for_unrelated_home_binding() {
        let registry = TriggerRegistry::new();
        let default_owner = Uuid::new_v4();
        let shop = std::sync::Arc::new(PausingProbe::default());
        registry
            .register_trigger_type(TriggerType::new_ns(
                crate::protocol::DEFAULT_NAMESPACE,
                "evt",
                "default",
                Box::new(std::sync::Arc::new(PausingProbe::default())),
                Some(default_owner),
            ))
            .await
            .unwrap();
        registry
            .register_trigger_type(TriggerType::new_ns(
                "shop",
                "evt",
                "shop",
                Box::new(std::sync::Arc::clone(&shop)),
                Some(Uuid::new_v4()),
            ))
            .await
            .unwrap();
        shop.pause.store(true, std::sync::atomic::Ordering::SeqCst);
        let binding = Trigger {
            home_namespace: "shop".into(),
            ..consumer_trigger("shop-binding", Uuid::new_v4())
        };
        let registration = registry.register_trigger(binding);
        tokio::pin!(registration);
        assert!(futures::poll!(&mut registration).is_pending());
        let cleanup = registry.unregister_worker(&default_owner);
        tokio::pin!(cleanup);
        let ready = futures::poll!(&mut cleanup).is_ready();
        shop.resume.notify_one();
        registration.await.unwrap();
        cleanup.await;
        assert!(ready, "DEFAULT provider disconnect blocked on an unrelated shop binding");
    }

    /// R10 — head-of-line blocking: an ownerless registration holds the gate
    /// while waiting for P's ack; P's read loop then handles UnregisterTrigger
    /// for the same id and blocks on that gate until the 10 s fail-open.
    #[tokio::test(start_paused = true)]
    async fn r10_read_loop_blocks_behind_ack_only_it_can_deliver() {
        let registry = std::sync::Arc::new(TriggerRegistry::new());
        let (p, _rx) = worker_provider(8);
        install_worker_provider(&registry, crate::protocol::DEFAULT_NAMESPACE, &p).await;
        let holder = {
            let registry = std::sync::Arc::clone(&registry);
            tokio::spawn(async move { registry.register_trigger(base_trigger("t")).await })
        };
        for _ in 0..4 {
            tokio::task::yield_now().await;
        }
        assert_eq!(p.pending_trigger_acks.len(), 1, "holder should be awaiting the ack");
        let start = tokio::time::Instant::now();
        let _ = registry.unregister_trigger("t".into(), None).await;
        let waited = tokio::time::Instant::now() - start;
        holder.await.unwrap().unwrap();
        assert!(
            waited < std::time::Duration::from_secs(1),
            "provider read loop stalled {waited:?} behind an ack only it can read"
        );
    }
}
```

</details>

<details>
<summary>Failure output on this branch</summary>

```
r1: unregister=Err(failed to send unregister trigger message through worker channel: channel closed) still_live=true provider_saw=["-t", "+t"]
r2: unregister=Err(… channel closed) still_parked=true replayed_on_return=["+t"] live_again=true
r3: unregister=Err(… channel closed) fallback_still_runs=true fallback_events=["+t", "-t", "+t"]
r4: same connection saw ["-t", "+t"]
r5: old provider still runs superseded binding: ["+t"]
r6: gate leaked with Some(1) retained deliveries
r7: after 5 restarts the binding retains Some(6) provider generations
r8: owner cleanup took 300s (virtual) for 30 bindings
r9: DEFAULT provider disconnect blocked on an unrelated shop binding
r10: provider read loop stalled 10s behind an ack only it can read
test result: FAILED. 0 passed; 10 failed
```

On `main` (`f2bc2b2cd`, portable module only): r1–r4 pass and r5 fails.

</details>
